draft/src/api.rs

```diff
@@ -1,16 +1,81 @@
 #[derive(Debug, Clone, Hash, Eq, PartialEq)]
 pub struct Card {
     pub name: String,
+    pub mana_cost: ManaCost,
 }
 
 impl Card {
-    pub fn new(name: &str) -> Self {
+    pub fn new(name: &str, mana_cost: ManaCost) -> Self {
         Self {
-           name: name.to_string(),
+            name: name.to_string(),
+            mana_cost,
         }
     }
 }
 
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum BasicMana{
+    Red,
+    Green,
+    Blue,
+    Black,
+    White,
+    // Can be fulfilled by any mana type including colorless.
+    Generic,
+    // Specifically requests colorless mana.
+    Colorless
```

Comment:
I don't think that Colorless should be part of BasicMana either, because it allows invalid states. A card can be Red AND Green, or any other color combination, but it can't be Colorless AND Green. So when noting what colors a card is, having Colorless as in BasicMana allows invalid states. This isn't just a theoretical concern either. If at any time you add a color to an already created card, for example, if it started as Colorless, now it is Colorless plus the new color, but it should JUST be the new color. If you need to represent the five colors PLUS colorless in a certain context, you can use the type Option<BasicMana>.

With this enum reduced down to just the five colors, it probably makes sense to rename it to just Color or ManaColor. Option<Color> having a value of None makes sense for colorless, but it isn't obvious what Option<BasicMana> being None means.

Reply:
Colorless mana is meant to represent the Eldrazi <> mana type, which feels like it should be its own BasicMana symbol since it cannot be derived from another type of data. My reasoning here is that the complex Mana type does not impose tight restrictions on your mana base the way the BasicMana type does. That said, I'm willing to go either way on it if you or anyone else feels strongly.